File: backend/app/models/schemas/dashboard_layout.py
"""Pydantic schemas for DashboardLayout entity."""

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


class DashboardLayoutCreate(BaseModel):
    """Schema for creating a new dashboard layout."""

    name: str = Field(..., description="Layout name", max_length=255)
    description: str | None = Field(None, description="Layout description")
    project_id: UUID | None = Field(None, description="Project scope (null = global)")
    is_template: bool = Field(False, description="Whether this layout is a reusable template")
    is_default: bool = Field(False, description="Whether this is the user's default layout for this scope")
    widgets: list[dict[str, object]] = Field(default_factory=list, description="Widget instances array")


class DashboardLayoutUpdate(BaseModel):
    """Schema for updating a dashboard layout. All fields optional."""

    name: str | None = Field(None, description="Layout name", max_length=255)
    description: str | None = Field(None, description="Layout description")
    is_template: bool | None = Field(None, description="Whether this layout is a reusable template")
    is_default: bool | None = Field(None, description="Whether this is the user's default layout")
    widgets: list[dict[str, object]] | None = Field(None, description="Widget instances array")


class DashboardLayoutRead(BaseModel):
    """Schema for reading dashboard layout data."""

    id: UUID
    name: str
    description: str | None
    user_id: UUID
    project_id: UUID | None
    is_template: bool
    is_default: bool
    widgets: list[dict[str, object]]
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class CloneTemplateRequest(BaseModel):
    """Schema for cloning a template layout."""

    project_id: UUID | None = Field(None, description="Project scope for the cloned layout")

File: frontend/src/types/dashboard-layout.ts
/**
 * Dashboard Layout Types
 *
 * TypeScript interfaces matching the backend Pydantic schemas defined in:
 * backend/app/models/schemas/dashboard_layout.py
 *
 * Non-versioned entity -- no EVCS, no time-travel params needed.
 */

/** Widget instance stored in a layout's widgets array. */
export interface WidgetConfig {
  instanceId: string;
  typeId: string;
  title?: string;
  config: Record<string, unknown>;
  layout: {
    x: number;
    y: number;
    w: number;
    h: number;
  };
}

/** Schema for creating a new dashboard layout. */
export interface DashboardLayoutCreate {
  name: string;
  description?: string | null;
  project_id?: string | null;
  is_template?: boolean;
  is_default?: boolean;
  widgets?: WidgetConfig[];
}

/** Schema for updating an existing dashboard layout. All fields optional. */
export interface DashboardLayoutUpdate {
  name?: string | null;
  description?: string | null;
  is_template?: boolean | null;
  is_default?: boolean | null;
  widgets?: WidgetConfig[] | null;
}

/**
 * Schema for reading dashboard layout data (API response).
 *
 * `role` / `scope` are seeder-only attributes on templates, exposed read-only so
 * the FE can resolve a user's role-tagged default template. They are never on
 * Create/Update (see backend schemas/dashboard_layout.py).
 */
export interface DashboardLayoutRead {
  id: string;
  name: string;
  description: string | null;
  user_id: string;
  project_id: string | null;
  is_template: boolean;
  is_default: boolean;
  widgets: WidgetConfig[];
  /** Template-only role tag (e.g. "cost-controller"); null = generic fallback. */
  role?: string | null;
  /** Template-only scope tag ("project" | "portfolio"); null when not a template. */
  scope?: string | null;
  created_at: string;
  updated_at: string;
}

/** Schema for cloning a template layout. */
export interface CloneTemplateRequest {
  project_id?: string | null;
  name?: string | null;
  /**
   * Whether the clone is the user's default for its scope (clearing any prior
   * default first). Used by the global dashboard's first-visit role-default
   * clone. See backend `clone_template(is_default=)`.
   */
  is_default?: boolean;
}
